docs/RationaleSeq.md

```diff
@@ -0,0 +1,89 @@
+of the output method (e.g. SystemVerilog).
+
+We have yet to flesh this dialect out to represent all common stateful
+components. We start with a register and will build from there.
+
+## Definitions
+
+For the sake of precision, we use the following definitions:
+
+- Physical devices:
+  - **Unclocked Latch:** A memory element which is only sensitive to the
+  levels of its inputs. Has no clock. Example: SR Latch.
+  - Clocked (gated) **latch:** A latch wherein the inputs are gated by a
+  clock. Transparent the entire time the clock is high. Generally referred to
+  as a "latch". Examples: "gated SR latch", "D latch".
+  - Edge-triggered **flip-flop:** An edge-sensitive memory element. Captures
+  the input value on one or both clock edges. Variants: posedge FF, negedge
+  FF, "edge-sensitive" FF (captures the input value on both edges),
+  resettable FF.
+- Abstract models:
+  - **Register:** A synchronous, resettable memory element. Can be
+  implemented using any of the above "circuit level" elements.
+
+## The register operation
+
+The `seq.reg` op models an abstract notion of a "register", independent of
+its implementation (e.g. latch, D flip-flop). It is intended to be 'lowered'
+to a specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic. It has
+three operands and one attribute:
+
+- **input**: The value to be captured at 'clock'. Generally called 'q'.
+Accepts any type, results in the same type.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A value which the state is set to upon reset. Optional.
+
+```mlir
+%d = seq.reg %input, %clk [, %reset [%resetValue] ] : $type(input)
+```
+
+### Rationale
+
+Several design decisions were made in defining this op:
+
+- Inclusion of 'reset' signal: the most common lowering case is expected to
+result in a SystemVerilog `always_ff`, which _should_ include an `if (rst)`
+block. This operand makes lowering to that form easier. Omission of if would
+require some (potentially complex) analysis to find the reset mux.
+- Inclusion of 'resetValue' attribute: if we have a 'reset' signal, we need
+to include a value to avoid complex inversion tricks.
+- Omission of 'enable': enables are easily modeled via a multiplexer on the
+input with one of the muxs inputs as the output of the register. This
+property makes 'enables' easier to detect than reset in the common case.
+- Omission of 'negedge' event on 'clock': this is easily modeled by inverting the
+clock value.
+- Omission of 'dual edge' event on 'clock': this is not terribly common. If
+we find a compelling use case, we could either create a new op for "dual
+edge" register, or create a "clock double" op which would interpose on the
+clock value. (And be detected by a lowering step).
+- Omission of edge conditions on 'reset': The reset style is generally a
```

Comment:
What about reset type? Are you assuming asynchronous reset always? That seems like a reasonable assumption (since synchronous reset can be pulled into the logic) unless you require a reset value

Reply:
I don't think it's important to specify a reset style at this level. Since this is the purely digital sequential logic layer, we don't want to have to take timing-dependent behavior into account.